src/internal/operators/shareReplay.ts

```diff
@@ -108,13 +108,17 @@ function shareReplayOperator<T>({
           subject.complete();
         },
       });
+      if (isComplete) {
+        subscription = undefined;
+      }
```

Comment:
I'm not sure how this solves a memory leak. Do we have tests or proof for this?

The *only* way this conditional passes is if the `source` completed synchronously. Which, if you [look right here](https://github.com/ReactiveX/rxjs/blob/1e1e022ed9fdc4a4bfed63cf151887031ac57398/src/internal/operators/shareReplay.ts#L106-L107), the `isComplete` is set to `true` and the `subscription` is set to `undefined` during that very same completion. 

In other words, this work has already been done by the time that `isComplete` would ever be `true`. So we can remove this statement entirely.

Reply:
> I'm not sure how this solves a memory leak. Do we have tests or proof for this?

Yes, we do, in [this link](https://stackblitz.com/edit/angular-7jmpmo?file=src%2Fapp%2Fvalue.service.ts). If you follow the reproduction steps of [this comment](https://github.com/ReactiveX/rxjs/issues/5931#issue-765399108), then you will notice that the memory leak is not present with these changes.

> The only way this conditional passes is if the source completed synchronously. 

Exactly!

> Which, if you look right here, the isComplete is set to true and the subscription is set to undefined during that very same completion.

That's what it _looks_ like. However, that's not what it actually happens... The thing is that when it completes synchronously, then in that moment `subscription` is `undefined`, because its value has not been assigned yet. It will be assigned after the callback gets executed. Therefore, the `if` statement is necessary to prevent the memory leak.

Reply:
Yeah. I was wrong. Sorry about that, I should have looked more carefully. Thank you for being diligent.

Here's what would have avoided my mistake: A detailed comment above this check like:

```ts
// Here we need to check to see if the source synchronously completed. Although
// we're setting `subscription = undefined` in the completion handler, if the source
// is synchronous, that will happen *before* subscription is set by the return of
// the `subscribe` call.
```

Reply:
No worries! I should have added a comment in the first place, since this is very counter intuitive.